src/dist/replication/lib/replica_learn.cpp

```diff
@@ -592,22 +592,31 @@ void replica::on_learn_reply(
             {
                 dinfo("%s: on_learn_reply[%016" PRIx64 "]: apply learned mutation %s", name(), req.signature, mu->name());
 
-                // write to shared log with no callback, the later 2pc ensures that logs
-                // are written to the disk
-                _stub->_log->append(mu, LPC_WRITE_REPLICATION_LOG, this, nullptr);
-
-                // because shared log are written without callback, need to manully
-                // set flag and write mutations to private log
-                mu->set_logged();
-                _private_log->append(mu, LPC_WRITE_REPLICATION_LOG, this, nullptr);                
-
-                // then we prepare
-                _prepare_list->prepare(mu, partition_status::PS_POTENTIAL_SECONDARY);
-
-                if (cache_range.first == 0 || mu->data.header.decree < cache_range.first)
-                    cache_range.first = mu->data.header.decree;
-                if (cache_range.second == 0 || mu->data.header.decree > cache_range.second)
-                    cache_range.second = mu->data.header.decree;
+                mutation_ptr existing_mutation = _prepare_list->get_mutation_by_decree(mu->data.header.decree);
+                if (existing_mutation!=nullptr && existing_mutation->data.header.ballot >= mu->data.header.ballot)
```

Comment:
the filter should be applied to prepare list only - the logs may be truncated already so logging is always necessary in this case.